usmqe/api/tendrlapi/common.py

```diff
@@ -1,19 +1,96 @@
-"""
-Tendrl REST API common functions.
-"""
+"""Tendrl REST API common functions."""
 
+import json
 import time
+
 import pytest
 import requests
+
 from usmqe.api.base import ApiBase
 
 LOGGER = pytest.get_logger("commonapi", module=True)
 
 
+class TendrlAuth(requests.auth.AuthBase):
+    """
+    Implementation of Tendrl Auth Method (Bearer Token) for requests
+    library, based on upstream documentation:
+
+    https://github.com/Tendrl/api/blob/master/docs/authentication.adoc
+
+    This way, auth. implementation is stored in a single place and it also
+    makes possible to omit auth object entrirely (thanks to design of
+    requests library) for testing purposes (eg. checks of corner cases to make
+    sure that security is not compromised) no matter if it makes sense from
+    Tendrl user perspective.
+
+    See also: http://docs.python-requests.org/en/master/user/authentication/
+    """
+
+    def __init__(self, token, username=None):
+        self._bearer_token = token
+        # metadata attributes for easier debugging, we need to trust login
+        # function to store correct values there
+        self.username = username
+
+    def __call__(self, r):
+        """Add Tendl Bearer Token into header of the request."""
+        headers = {
+            "Authorization": "Bearer {}".format(self._bearer_token),
+            }
+        r.prepare_headers(headers)
+        return r
+
+
+def login(username, password, asserts_in=None):
+    """
+    Login Tendrl user.
+
+    Args:
+        username: name of user that is going logged in
+        password: password for username
+        asserts_in: assert values for this call and this method
+
+    Returns requests auth object (instance of TendrlAuth)
+    """
+    pattern = "login"
+    post_data = {"username": username, "password": password}
+    request = requests.post(
+        pytest.config.getini("usm_api_url") + pattern,
+        data=json.dumps(post_data))
+    ApiBase.print_req_info(request)
+    ApiBase.check_response(request, asserts_in)
+    token = request.json().get("access_token")
+    LOGGER.info("access_token: {}".format(token))
+    auth = TendrlAuth(token, username)
+    return auth
+
+
+def logout(auth, asserts_in=None):
+    """
+    Logout Tendrl user.
+
+    Args:
+        asserts_in: assert values for this call and this method
+        auth: TendrlAuth object (defines bearer token header)
+    """
+    pattern = "logout"
+    request = requests.delete(
+        pytest.config.getini("usm_api_url") + pattern,
+        auth=auth)
+    ApiBase.print_req_info(request)
+    ApiBase.check_response(request, asserts_in)
```

Comment:
logout method could and probably should be part of TendrlApi class, for login it makes sense to have it outside though. However I agree that it could look strange in test, but to have it in the class makes sense as login is really only call which doesn't require authentication.